cnxepub/formatters.py

```diff
@@ -426,7 +429,77 @@ def exercise_callback_factory(match, url_template,
     """Create a callback function to replace an exercise by fetching from
     a server."""
 
-    def _replace_exercises(elem):
+    def _annotate_exercise(elem, exercise, page_uuids):
+        """Annotate exercise based upon tag data"""
+        tags = exercise['items'][0].get('tags')
+        if not tags:
+            return
+
+        # Annotate exercise with required context, if it exists
+        modules, feature = [], ''
+        for tag in tags:
+            if 'context-cnxmod:' in tag:
+                modules.append(tag.split(':')[1])
+            if 'context-cnxfeature:' in tag:
+                # Note: Assuming this tag will never be duplicated with
+                # multiple values in the exercise data
+                feature = tag.split(':')[1]
+
+        # It's possible that the feature tag is not present, or present but
+        # not valid (e.g. value of "context-cnxfeature:").
+        if not feature:
+            return
+
+        # There may be multiple `context-cnxmod:{uuid}` tags, each of which
+        # could be the parent page for this exercise, a different page in this
+        # book, or even invalid altogether. We'll prefer the first, fallback
+        # to the second, and not annotate in the last case.
+
+        # Determine parent page for this exercise
+        parent_page_uuid = None
+        parent_page_elem = None
+        for parent in elem.iterancestors():
```

Comment:
Yeah, let me try using `find` here.

Just FYI - I was having having seg fault issues when I tried using `xpath`. I think these occurred due to thread safety issues, but [from here](https://lxml.de/FAQ.html#can-i-use-threads-to-concurrently-access-the-lxml-api) it wasn't obvious to me whether this would be expected or not 🤷‍♂️ . Curious if you've had experience with this type of problem with threads and `lxml` in the past?

Reply:
@philschatz - Actually, I remember now why I went this route in the first place. I seemed to me like `find()` only supports searching down the tree (I'm not really clear on the subset of xpath supported by `ElementPath` TBH) and I hit the `xpath` seg fault issues with that approach. Can you point me to the appropriate incantation / docs if I'm wrong about the `find()` limitation 😅 ? Otherwise, perhaps I can take a deeper look at the `xpath` segfaults to see if I can ferret out the root cause.

Reply:
Actually, never mind. [This change](https://github.com/openstax/cnx-epub/pull/177/commits/f8c2c5ce92f9a1dbbbd7d310bd04103fc7bd83d3) seems to work fine using `xpath`. I think the query I was doing when I saw seg faults was down tree, but I tested this with real books and it doesn't seem to keel over.

Reply:
Oh my thanks for digging through to find the root of the multithreaded problem and fixing it (using both `.xpath()` and `.find()`